### PR TITLE
Add the function to convert column of floating points with `nan`s into `bitmask`

### LINKED ISSUE #1788: [FEA] Function to convert column of floating points with NaNs into bitmask
It is necessary to be able to take a column of floating point values, potentially containing NaNs, and generate a bitmask where every NaN is indicated as null by the bitmask.

The function would look something like:
```
bit_mask_t * nans_to_nulls(gdf_column const * col);
```

`col` may be nullable and must be either `GDF_FLOAT32` or `GDF_FLOAT64`. The returned bitmask will be copied from `col` and updated to have a 0 at bit `i` if `col->data[i]` is `NaN`.

This should be very easy and efficient to implement in a kernel that uses [`isnan`](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html) and [`__ballot_sync` ](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-vote-functions).

This would be used to replace an inefficient Numba function that is currently doing the same thing: https://github.com/rapidsai/cudf/blob/branch-0.8/python/cudf/utils/cudautils.py#L362. Once the libcudf primitive is created, usage of `mask_from_devary` should be replaced with `nans_to_nulls`.

### PR DESCRIPTION
**Pull request by hummingtree**
This PR adds a function that checks the elements of a column of floating point numbers (`GDF_FLOAT32` and `GDF_FLOAT64`)  and mark bits in `bitmask` as null if the corresponding element is a `nan`. The resulting `null_count` is also calculated in the same kernel and returned as the `second` of the returned `std::pair` object.
```
std::pair<bit_mask::bit_mask_t*, gdf_size_type> nans_to_nulls(gdf_column const& input);
```

Closes #1788

### CODE REVIEW

File: cpp/src/utilities/column_utils.cuh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    bit_index += blockDim.x * gridDim.x;
+    active_threads =
+        __ballot_sync(active_threads, bit_index < num_bits);
+  }
+
+}
+
+template <typename BitContainer, typename Predicate, typename Size>
+BitContainer* null_if(
```
Comment (harrism) [inline cpp/src/utilities/column_utils.cuh]: Personally I would prefer a `valid_if` function (I like "default true" semantics), with reversed predicate logic.

File: cpp/src/utilities/column_utils.cuh [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  Size bit_index = threadIdx.x + blockIdx.x * blockDim.x;
+
+  BitContainer active_threads =
+      __ballot_sync(0xffffffff, bit_index < num_bits);
+
+  while (bit_index < num_bits) {
+    
+    bool const predicate_is_true = p(bit_index);
+    // This function name is `null_if` so if predicate is true the bit is to be set to off
```
Comment (harrism) [inline cpp/src/utilities/column_utils.cuh]: See my comment about `valid_if`, so you don't have to negate the predicate. Negating the predicate is confusing...

File: cpp/src/utilities/column_utils.cuh [hunk 3]
```diff
@@ -0,0 +1,96 @@
+/*
```
Comment (harrism) [inline cpp/src/utilities/column_utils.cuh]: I don't like "hiding" this functionality in a file called "column_utils". I think you should create a file called `valid_if` (or `null_if`) if you keep that name. Put it in `src/bitmask`.

File: cpp/src/utilities/column_utils.cu [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  predicate_is_nan() = delete;
+  
+  predicate_is_nan(const gdf_column input_): input(input_) {}
+
+};
+
+
+} // namespace cudf
+
+bit_mask_t* nans_to_nulls(gdf_column const* col){
```
Comment (harrism) [inline cpp/src/utilities/column_utils.cu]: This should be in `namespace cudf`.

File: cpp/tests/binaryop/integration/assert-binops.h [hunk 5]
```diff
@@ -148,5 +148,5 @@ void ASSERT_BINOP(cudf::test::column_wrapper<TypeOut>& out,
 } // namespace binop
 } // namespace test
 } // namespace cudf
-
+//
```
Comment (harrism) [inline cpp/tests/binaryop/integration/assert-binops.h]: Revert this change (assume it was accidental?).

File: cpp/src/utilities/column_utils.cu [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  
+  const bit_mask_t* source_mask = reinterpret_cast<bit_mask_t*>(col->valid);
```
Comment (harrism) [inline cpp/src/utilities/column_utils.cu]: Need to handle empty columns. Should probably just return an nullptr if the input column has zero size.

File: cpp/tests/bitmask/nans_to_nulls_tests.cu [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+using bit_mask::bit_mask_t;
+
+template <typename T>
+struct BitMaskTest : GdfTest {};
+
+using test_types =
+    ::testing::Types<float, double>;
+TYPED_TEST_CASE(BitMaskTest, test_types);
+
```
Comment (harrism) [inline cpp/tests/bitmask/nans_to_nulls_tests.cu]: Need a test of empty input column, as well as error conditions (unsupported type).

File: cpp/src/utilities/column_utils.cu [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */ 
+
+#include <utilities/column_utils.cuh>
+
+namespace cudf {
+
+template <typename T>
```
Comment (harrism) [inline cpp/src/utilities/column_utils.cu]: This should be in `namespace cudf::detail`, not top-level cudf namespace.

File: cpp/include/cudf/functions.h [hunk 9]
```diff
@@ -845,3 +845,6 @@ gdf_error gdf_from_dlpack(gdf_column** columns,
 gdf_error gdf_to_dlpack(DLManagedTensor_ *tensor,
                         gdf_column const * const * columns,
                         gdf_size_type num_columns);
+
+
+bit_mask::bit_mask_t* nans_to_nulls(gdf_column const* col);
```
Comment (harrism) [inline cpp/include/cudf/functions.h]: I think this is a transform. So it should probably go in the same header as cudf::transform, not in something called `column_utils`. This is an API function, not a utility. As such, it should be in namespace cudf.

Also, this needs documentation and Cython bindings.

File: cpp/src/utilities/column_utils.cuh (hunk 2 above)
Comment (jrhemstad) [reply cpp/src/utilities/column_utils.cuh]: Yeah, that's true. I originally asked for `null_if`, but I got my logic mixed up and agree that `valid_if` is a better name.

File: cpp/src/utilities/column_utils.cuh (hunk 1 above)
Comment (jrhemstad) [inline cpp/src/utilities/column_utils.cuh]: I realized that this function should also return the `null_count`.

File: cpp/src/utilities/column_utils.cuh [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+#include <bitmask/legacy/bit_mask.cuh>
+
+using bit_mask::bit_mask_t;
+
+namespace cudf {
+
+constexpr int warp_size = 32;
+
+template <bool source_mask_valid, typename BitContainer, typename Predicate, typename Size>
+__global__ void null_if_kernel(
```
Comment (jrhemstad) [inline cpp/src/utilities/column_utils.cuh]: Should also add computing the null count.

File: cpp/include/cudf/functions.h (hunk 9 above)
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: Return a `std::pair<bit_mask_t*, gdf_size_type>` for the mask and the null count.

Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: Inputs are pass by const ref.
```suggestion
bit_mask::bit_mask_t* nans_to_nulls(gdf_column const& col);
```

File: cpp/src/utilities/column_utils.cu [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  
+  const bit_mask_t* source_mask = reinterpret_cast<bit_mask_t*>(col->valid);
+  
+  switch(col->dtype){
+    case GDF_FLOAT32:
+      return cudf::null_if(source_mask, cudf::predicate_is_nan<float>(*col), col->size);
+    case GDF_FLOAT64:
+      return cudf::null_if(source_mask, cudf::predicate_is_nan<double>(*col), col->size);
+    default:
+      CUDF_EXPECTS(false, "Unsupported data type for is_nan()");
```
Comment (jrhemstad) [inline cpp/src/utilities/column_utils.cu]:
```suggestion
      CUDF_FAIL("Unsupported data type for is_nan()");
```

**Comment by hummingtree** [comment]
rerun tests

File: cpp/src/bitmask/valid_if.cuh [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+  // reduction
+  if (lane_id == lane) {
+    smem[warp_id] = __popc(bit_mask);
+  }
+  __syncthreads();
+
+  if (warp_id == 0) {
+    // Here I am assuming maximum block size is 1024 and 1024 / 32 = 32
+    // so one single warp is enough to do the reduction over different warps
+    bit_mask = (lane_id < (blockDim.x / warp_size)) ? smem[lane_id] : 0;
```
Comment (harrism) [inline cpp/src/bitmask/valid_if.cuh]: Why is this called bit_mask? `smem[lane_id]` contains the result of __popc(), which is a count, not a mask.  I think you should call this `count`...
```suggestion
    gdf_size_type count = (lane_id < (blockDim.x / warp_size)) ? warp_counts[lane_id] : 0;
```

File: cpp/src/bitmask/valid_if.cuh [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+/**
+ * @brief for each warp in the block do a reduction (summation) of the
+ * `__popc(bit_mask)` on a certain lane (default is lane 0).
+ * @param[in] bit_mask The bit_mask to be reduced.
+ * @return[out] result of each block is returned in thread 0.
+ */
+template <class bit_container, int lane = 0>
+__device__ __inline__ gdf_size_type single_lane_popc_block_reduce(bit_container bit_mask) {
+  
+  static __shared__ gdf_size_type smem[warp_size];
```
Comment (harrism) [inline cpp/src/bitmask/valid_if.cuh]: `smem` is a generic name, but this memory is used for only one thing: warp counts, so make the name reflect the contents.

```suggestion
  static __shared__ gdf_size_type warp_counts[warp_size];
```

File: cpp/src/bitmask/valid_if.cuh [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    // The cub::warpReduce could be replaced with a __shfl_down_sync?
+    /**
+      #pragma unroll
+      for(int offset = warp_size/2; offset > 0; offset /= 2){
+        bit_mask += __shfl_down_sync(0xffffffffu, bit_mask, offset);
+      }
+    */
+    __shared__
+        typename cub::WarpReduce<gdf_size_type>::TempStorage temp_storage;
+    bit_mask = cub::WarpReduce<gdf_size_type>(temp_storage).Sum(bit_mask);
```
Comment (harrism) [inline cpp/src/bitmask/valid_if.cuh]:
```suggestion
    count = cub::WarpReduce<gdf_size_type>(temp_storage).Sum(count);
```

File: cpp/src/bitmask/valid_if.cuh [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  return bit_mask;
```
Comment (harrism) [inline cpp/src/bitmask/valid_if.cuh]:
```suggestion
  return count;
```

File: cpp/src/bitmask/valid_if.cuh [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+  
+  static __shared__ gdf_size_type smem[warp_size];
+  
+  int lane_id = (threadIdx.x % warp_size);
+  int warp_id = (threadIdx.x / warp_size);
+
+  // Assuming one lane of each warp holds the value that we want to perform
+  // reduction
+  if (lane_id == lane) {
+    smem[warp_id] = __popc(bit_mask);
```
Comment (harrism) [inline cpp/src/bitmask/valid_if.cuh]:
```suggestion
    warp_counts[warp_id] = __popc(bit_mask);
```

File: python/cudf/cudf/bindings/utils.pxd [hunk 17]
```diff
@@ -5,9 +5,18 @@
 # cython: embedsignature = True
```
[the remark this answers is not in the document]
Comment (hummingtree) [reply python/cudf/cudf/bindings/utils.pxd]: @shwina @kkraus14 What should be the right place for the binding here?

[the remark this answers is not in the document]
Comment (hummingtree) [reply python/cudf/cudf/bindings/utils.pxd]: @harrism The bindings for `cudf::transform` are in the `unaryops.*` files ...

[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/bindings/utils.pxd]: Right now this is OK. Maybe we can address this in https://github.com/rapidsai/cudf/issues/1983
